File: ContactsManagerAppWithJavaROOMDBMVVM/app/src/main/java/com/masterclass/contactsmanagerappwithjavaroomdbmvvm/repositories/ContactsRepository.java
package com.masterclass.contactsmanagerappwithjavaroomdbmvvm.repositories;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;
import androidx.lifecycle.LiveData;
import com.masterclass.contactsmanagerappwithjavaroomdbmvvm.database.ContactDAO;
import com.masterclass.contactsmanagerappwithjavaroomdbmvvm.database.ContactDatabase;
import com.masterclass.contactsmanagerappwithjavaroomdbmvvm.database.Contacts;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ContactsRepository {
    //Available data sources: RoomDb
    //Execute all methods from the DAO
    private final ContactDAO contactDAO;

    //To avoid app hanging/app crash, we do this task in the background service/ background thread
    //creates a different thread/Threads separation
    ExecutorService executorService;

    //Post results to main thread/Updating the Ui/
    Handler handler;

    public ContactsRepository(Application application) {
        ContactDatabase contactDatabase = ContactDatabase.getInstance(application);
        this.contactDAO = contactDatabase.getContactDAO();
        executorService = Executors.newSingleThreadExecutor();
        handler = new Handler(Looper.getMainLooper());
    }

    public void addContact(Contacts contact){

        //Execute the task in a separate thread using runnable
        executorService.execute(new Runnable() {
            @Override
            public void run() {
                //code is executed asynchronously in a separate thread
                contactDAO.insert(contact);
            }
        });
    }

    public void deleteContact(Contacts contact){
        executorService.execute(new Runnable() {
            @Override
            public void run() {
                contactDAO.delete(contact);
            }
        });
    }

    public LiveData<List<Contacts>> getAllContacts(){
        return contactDAO.getAllContacts();
    }


}

File: ContactsManagerAppWithKotlinROOMDBMVVM2/app/src/main/java/com/masterclass/contactsmanagerappwithkotlinroomdbmvvm/repositories/ContactsRepository.kt
package com.masterclass.contactsmanagerappwithkotlinroomdbmvvm.repositories

import android.annotation.SuppressLint
import android.app.Application
import android.os.Handler
import android.os.Looper
import androidx.lifecycle.LiveData
import com.masterclass.contactsmanagerappwithkotlinroomdbmvvm.database.ContactDAO
import com.masterclass.contactsmanagerappwithkotlinroomdbmvvm.database.ContactDatabase
import com.masterclass.contactsmanagerappwithkotlinroomdbmvvm.database.Contacts
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ContactsRepository(application: Application) {

    private val contactDAO: ContactDAO
    private val executorService: ExecutorService
    private val handler: Handler

    init {
        val contactDatabase = ContactDatabase.getInstance(application)
        contactDAO = contactDatabase.getContactDAO()!!
        executorService = Executors.newSingleThreadExecutor()
        handler = Handler(Looper.getMainLooper())
    }

    fun addContact(contact: Contacts) {
        executorService.execute {
            // code is executed asynchronously in a separate thread
            contactDAO.insert(contact)
        }
    }

    fun deleteContact(contact: Contacts) {
        executorService.execute {
            contactDAO.delete(contact)
        }
    }

    fun getAllContacts(): LiveData<List<Contacts>> {
        return contactDAO.getAllContacts()
    }
}